File: composeApp/src/commonMain/kotlin/dev/mks/duskread/ui/theme/Tokens.kt
package dev.mks.duskread.ui.theme

import androidx.compose.ui.unit.dp

/*
 * Design tokens: the values that carry a decision rather than a measurement.
 *
 * These are plain objects rather than CompositionLocals because none of them
 * vary by theme, platform or window size — a card is 16dp round in dark mode
 * too. Colour and typography, which *do* vary, live in [DuskReadTheme] and
 * are reached through `MaterialTheme`.
 *
 * **What belongs here:** a value used in more than one place, or one whose
 * exact number matters to how the app reads. **What does not:** optical
 * one-offs. Padding of
 * `top = 14.dp, bottom = 13.dp` is a nudge to make a specific label sit right,
 * not a rule, and naming it would imply a system that is not there.
 *
 * Spacing follows a 2dp rhythm. New values should land on it.
 *
 * The numbers themselves have moved to [DesignTokens], which carries no
 * Compose types and so can be read from Swift. These objects stay because
 * they are what the Compose UI reads — this file is now the `Dp` face of the
 * same values, not a second set of them.
 */

/** Sizes that shape the page rather than decorate it. */
object Layout {
    /** Horizontal padding for reading surfaces. */
    val ReadingGutter = DesignTokens.ReadingGutter.dp

    /** Horizontal padding for list surfaces, which carry their own card insets. */
    val ListGutter = DesignTokens.ListGutter.dp

    /** Bottom inset so the last item clears the floating bar. */
    val BarClearance = DesignTokens.BarClearance.dp

    /** Every face of the floating bar is this tall; only the width changes between them. */
    val BarHeight = DesignTokens.BarHeight.dp

    /**
     * What the floating bar keeps between itself and the safe area.
     */
    val BarInset = DesignTokens.BarInset.dp

    /**
     * The one width that changes the plan: below it the floating bar, above it the rail.
     */
    val TwoPaneBreakpoint = DesignTokens.TwoPaneBreakpoint.dp

    /**
     * The vertical navigation rail replacing the floating bar when wide. Narrow enough to
     * read as an edge rather than a column of its own.
     */
    val RailWidth = DesignTokens.RailWidth.dp

    /**
     * The widest a column of prose is allowed to get, regardless of how much room there
     * is around it.
     */
    val ReadingMeasure = DesignTokens.ReadingMeasure.dp

    /** [ListGutter], opened up once there is room. */
    val WideListGutter = DesignTokens.WideListGutter.dp
}

/** Corner radii, largest to smallest. */
object Radius {
    /**
     * Dashboard and list cards.
     */
    val Card = DesignTokens.RadiusCard.dp

    /**
     * Everything that sits inside or below a card: the filled call-to-action, text
     * fields, dashboard and list rows, the timer's state chips.
     */
    val Inline = DesignTokens.RadiusInline.dp

    /**
     * The mockup's `.pill` and `.sourcechip`: 3dp, which at these sizes is a softened
     * corner rather than a rounded one.
     */
    val Chip = DesignTokens.RadiusChip.dp
}

/** Line weights. */
object Stroke {
    val Hairline = DesignTokens.StrokeHairline.dp
}

/** Gaps that recur. Anything used once stays a literal at its call site. */
object Space {
    /** Between chips in a row. */
    val ChipGap = DesignTokens.ChipGap.dp

    /** Between cards in a list. */
    val CardGap = DesignTokens.CardGap.dp
}

/**
 * Durations, in milliseconds.
 */
object Motion {
    /** Pushing to a full-screen destination: the incoming screen slides and fades in. */
    const val PushIn = DesignTokens.MotionPushIn

    /** …fading slightly slower than it slides, so it does not vanish mid-travel. */
    const val PopFade = DesignTokens.MotionPopFade

    /** Cross-fades that should not draw attention: tab and pane swaps. */
    const val Fade = DesignTokens.MotionFade

    /** Chip and bar state changes. */
    const val Chip = DesignTokens.MotionChip
}

File: iosApp/iosApp/Design/Tokens.swift
import ComposeApp
import SwiftUI

/// Layout, radius, stroke and spacing, read from the shared module.
///
/// These are thin because they have to be: the moment a number is typed on
/// this side rather than read from `DesignTokens`, the two UIs have started
/// drifting. Kotlin's `dp` and SwiftUI's points are the same unit at the same
/// scale, so there is nothing to convert.
enum Layout {
    static let readingGutter = CGFloat(DesignTokens.shared.ReadingGutter)
    static let listGutter = CGFloat(DesignTokens.shared.ListGutter)
    static let wideListGutter = CGFloat(DesignTokens.shared.WideListGutter)
    static let barClearance = CGFloat(DesignTokens.shared.BarClearance)
    static let barHeight = CGFloat(DesignTokens.shared.BarHeight)
    static let barInset = CGFloat(DesignTokens.shared.BarInset)
    static let twoPaneBreakpoint = CGFloat(DesignTokens.shared.TwoPaneBreakpoint)
    static let railWidth = CGFloat(DesignTokens.shared.RailWidth)
    static let readingMeasure = CGFloat(DesignTokens.shared.ReadingMeasure)
}

enum Radius {
    static let card = CGFloat(DesignTokens.shared.RadiusCard)
    static let inline = CGFloat(DesignTokens.shared.RadiusInline)
    /// The one to reach for: a softened corner, not a round one.
    static let chip = CGFloat(DesignTokens.shared.RadiusChip)
}

enum Stroke {
    /// One **point**, not one physical pixel — the hairline is a weight, not a
    /// hardware detail, and `1 / UIScreen.scale` would make it vanish.
    static let hairline = CGFloat(DesignTokens.shared.StrokeHairline)
}

enum Space {
    static let chipGap = CGFloat(DesignTokens.shared.ChipGap)
    static let cardGap = CGFloat(DesignTokens.shared.CardGap)
}

/// Durations, mirrored as timing curves rather than springs.
///
/// The shared UI has four `tween`s on Compose's default easing and not one
/// spring anywhere. SwiftUI's idiom is springs, and reaching for one here
/// would be redesigning the motion rather than porting it — so this is the
/// same cubic Compose uses, at the same four durations.
enum Motion {
    static let pushIn = seconds(DesignTokens.shared.MotionPushIn)
    static let popFade = seconds(DesignTokens.shared.MotionPopFade)
    static let fade = seconds(DesignTokens.shared.MotionFade)
    static let chip = seconds(DesignTokens.shared.MotionChip)

    private static func seconds(_ ms: Int32) -> Double { Double(ms) / 1000 }

    /// Compose's `FastOutSlowInEasing`.
    static func ease(_ duration: Double) -> Animation {
        .timingCurve(0.4, 0.0, 0.2, 1.0, duration: duration)
    }
}
